### PR TITLE
Build minimal Docker image that runs without root priviledges

### PR DESCRIPTION
**Pull request by regel**
Running images as UID 0 is avoided in Openshift and K8s. I've made changes to the Dockerfile to build a minimal image that complies with security requirements in these platforms.

It's also possible to override the two ARG at the start with docker --build-arg flag if you want to publish images for other architectures.

I did not edit .github/workflows/docker. I will edit another PR with suggestions on that.

### CODE REVIEW

**Comment by prologic** [comment]
Nice 👌

**Review by prologic — APPROVED** [review]

**Comment by prologic** [comment]
Why 65532?

And what's distroless?

File: Dockerfile [hunk 1]
```diff
@@ -1,14 +1,37 @@
...
+RUN CGO_ENABLED=0 GOOS=${GOOS} GOARCH=${GOARCH} GO111MODULE=on go build -a -o bitraft
 
-# Runtime
-FROM alpine
+RUN mkdir data
 
-COPY --from=build /src/bitraft /bitraft
+# Use distroless as minimal base image
+# Refer to https://github.com/GoogleContainerTools/distroless for more details
+FROM gcr.io/distroless/static:nonroot
```
Comment (prologic) [inline Dockerfile:25]: Can we just use `scratch` here or `alpine`?

After having read about distroless I _really_ don't think this adds much value here at all.

**Review by prologic — CHANGES_REQUESTED** [review]

**Comment by regel** [comment]
> Why 65532?

The USER instruction sets the default user name (or UID) to use when running the image and for any RUN, CMD and ENTRYPOINT instructions that follow it in the Dockerfile, unless the -u option is specified at run time.

Some people eg Bitnami use 1001 as a convention to package all their Docker images and remove 'root' (UID 0) permissions. The idea is to pick a number !=0 and >1000. 1000 is generally the first UID assigned to a user on the host, so 1001 is intended to avoid UID collisions between host and container.

Using 65532 as max value is a similar convention as 1001, with the extra addition in distroless images of the /etc/passwd file that declares this UID.

Here since persistent volumes are required to store data users would run docker run ... -v $PWD/data:/data -u $(id -u):$(id -g) to avoid file permissions issues when sharing volumes between host and container.

> And what's distroless?

Similar to alpine. Some people prefer distroless for security reasons when building minimal images.

Using the 'debug' tag in distroless images is adding a shell. Below is the illustration of the /etc/passwd file.

$ docker run -ti gcr.io/distroless/base-debian10:debug --entrypoint=sh
/ # cat /etc/passwd 
root:x:0:0:root:/root:/sbin/nologin
nobody:x:65534:65534:nobody:/nonexistent:/sbin/nologin
nonroot:x:65532:65532:nonroot:/home/nonroot:/sbin/nologin


Choosing between alpine or distroless as a base image alternative is easily achieved via an ARG at the top of the Dockerfile. The more freedom, the better. I can push one more simple commit in this PR to change the default value.

Usually, CI workflows will build all different permutations of architecture (amd64, arm64....) and base images and publish them with a suffix in their tag name, eg 1.1-alpine, 1.1-distroless, 1.1-buster, and so on. The team at Docker who is publishing the official images is using this tag name convention, and since then people seem to follow this practice but again it's just a convention.

**Comment by regel** [comment]
Commit added with Alpine as default base image. I used a build-arg so it's flexible and easy to select another base image without changing Dockerfile.

**Comment by prologic** [comment]
Thanks for the explanation!

**Pull request merged**